pallets/balances/src/lib.rs

```diff
@@ -462,6 +462,22 @@ decl_module! {
             let dest = T::Lookup::lookup(dest)?;
             Self::transfer_core(&source, &dest, value, None, ExistenceRequirement::AllowDeath)?;
         }
+
+        /// Burns the given amount of tokens from the caller's free balance.
+        #[weight = SimpleDispatchInfo::FixedNormal(200_000)]
+        pub fn burn_free_balance(origin, amount: T::Balance) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            let total_issuance = Self::total_issuance();
+            let amount = cmp::min(total_issuance, cmp::min(amount, Self::account(&who).free));
```

Comment:
Instead of burning max allowed, let's throw an error if the user tries to burn more than allowed.